File: backend/app/schemas/analysis.py
import uuid
from datetime import datetime

from pydantic import BaseModel

from app.models.analysis import STAGE_ORDER, Analysis, AnalysisStatus

StageState = str  # "complete" | "processing" | "pending"


class AnalysisResponse(BaseModel):
    id: uuid.UUID
    dataset_id: uuid.UUID
    dataset_version_id: uuid.UUID
    status: AnalysisStatus
    progress: int
    stages: dict[str, StageState]
    error: str | None
    started_at: datetime | None
    completed_at: datetime | None
    pipeline_version: int
    prompt_version: int
    retry_count: int
    data_quality: dict | None
    ai_findings: dict | None
    recommendations: dict | None

    model_config = {"from_attributes": True}

    @classmethod
    def from_analysis(cls, analysis: Analysis) -> "AnalysisResponse":
        stages: dict[str, StageState] = {}
        progress = 0

        if analysis.status == AnalysisStatus.READY:
            stages = {s.value: "complete" for s in STAGE_ORDER}
            progress = 100
        elif analysis.status == AnalysisStatus.FAILED:
            # Granular per-stage state isn't reconstructable once terminal -- the error string
            # (prefixed with the failing stage) carries that context instead.
            stages = {}
        else:
            current_index = STAGE_ORDER.index(analysis.status)
            stages = {
                s.value: (
                    "complete"
                    if i < current_index
                    else "processing"
                    if i == current_index
                    else "pending"
                )
                for i, s in enumerate(STAGE_ORDER)
            }
            progress = round(current_index / len(STAGE_ORDER) * 100)

        return cls(
            id=analysis.id,
            dataset_id=analysis.dataset_id,
            dataset_version_id=analysis.dataset_version_id,
            status=analysis.status,
            progress=progress,
            stages=stages,
            error=analysis.error,
            started_at=analysis.started_at,
            completed_at=analysis.completed_at,
            pipeline_version=analysis.pipeline_version,
            prompt_version=analysis.prompt_version,
            retry_count=analysis.retry_count,
            data_quality=analysis.data_quality or None,
            ai_findings=analysis.ai_findings or None,
            recommendations=analysis.recommendations or None,
        )

File: frontend/src/lib/api/analysis.ts
import { apiClient } from "@/lib/api/client";
import type { VisualizationRecommendation } from "@/lib/api/types";

/** Mirrors backend/app/schemas/analysis.py. Kept by hand until codegen exists. */

export type AnalysisStageStatus = "complete" | "processing" | "pending";

export type AnalysisStatus =
  | "queued"
  | "profiling_quality"
  | "building_ai_context"
  | "ai_analyzing"
  | "generating_recommendations"
  | "validating"
  | "ranking"
  | "generating_previews"
  | "ready"
  | "failed";

export interface DataQualityIssue {
  type: string;
  column: string | null;
  description: string;
  severity: "low" | "medium" | "high";
  recommendation: string;
}

export interface DataQuality {
  score: number;
  issues: DataQualityIssue[];
}

export interface AnalysisRecommendations {
  top: VisualizationRecommendation[];
  shortfall_reason: string | null;
}

export interface Analysis {
  id: string;
  dataset_id: string;
  dataset_version_id: string;
  status: AnalysisStatus;
  progress: number;
  stages: Record<string, AnalysisStageStatus>;
  error: string | null;
  started_at: string | null;
  completed_at: string | null;
  pipeline_version: number;
  prompt_version: number;
  retry_count: number;
  data_quality: DataQuality | null;
  ai_findings: Record<string, unknown> | null;
  recommendations: AnalysisRecommendations | null;
}

/** Uploading a dataset alone starts the whole pipeline server-side -- this is the only
 * endpoint the frontend needs to reach the final recommendations; no manual "Analyze" call. */
export function getAnalysis(datasetId: string) {
  return apiClient.get<Analysis>(`/api/datasets/${datasetId}/analysis`);
}

export function retryAnalysis(datasetId: string) {
  return apiClient.post<Analysis>(`/api/datasets/${datasetId}/analysis/retry`);
}

export const ANALYSIS_STAGE_LABELS: Record<string, string> = {
  queued: "Queued",
  profiling_quality: "Analyzing data quality",
  building_ai_context: "Building AI context",
  ai_analyzing: "AI is analyzing the dataset",
  generating_recommendations: "Generating recommendations",
  validating: "Validating visualizations",
  ranking: "Ranking visualizations",
  generating_previews: "Generating previews",
};
